routing/routing_tests/index_graph_tools.cpp

```diff
@@ -44,27 +44,70 @@ AStarAlgorithm<IndexGraphStarter>::Result CalculateRoute(IndexGraphStarter & sta
   RoutingResult<Joint::Id> routingResult;
+
   auto const resultCode = algorithm.FindPath(
       starter, starter.GetStartJoint(), starter.GetFinishJoint(), routingResult, {}, {});
 
   starter.RedressRoute(routingResult.path, roadPoints);
   return resultCode;
 }
 
+void TestRouteSegments(IndexGraphStarter & starter,
+                       AStarAlgorithm<IndexGraphStarter>::Result expectedRouteResult,
+                       vector<RoadPoint> const & expectedRoute)
+{
+  vector<RoadPoint> route;
+  auto const resultCode = CalculateRoute(starter, route);
+  TEST_EQUAL(resultCode, expectedRouteResult, ());
+  TEST_EQUAL(route, expectedRoute, ());
+}
+
 void TestRouteGeometry(IndexGraphStarter & starter,
                        AStarAlgorithm<IndexGraphStarter>::Result expectedRouteResult,
                        vector<m2::PointD> const & expectedRouteGeom)
 {
   vector<RoadPoint> route;
   auto const resultCode = CalculateRoute(starter, route);
+
   TEST_EQUAL(resultCode, expectedRouteResult, ());
-  TEST_EQUAL(route.size(), expectedRouteGeom.size(), ());
+  vector<m2::PointD> geom;
   for (size_t i = 0; i < route.size(); ++i)
   {
-    // When PR with applying restricions is merged IndexGraph::GetRoad() should be used here instead.
-    RoadGeometry roadGeom = starter.GetGraph().GetGeometry().GetRoad(route[i].GetFeatureId());
-    CHECK_LESS(route[i].GetPointId(), roadGeom.GetPointsCount(), ());
-    TEST_EQUAL(expectedRouteGeom[i], roadGeom.GetPoint(route[i].GetPointId()), ());
+    m2::PointD const & pnt = starter.GetPoint(route[i]);
+    if (geom.empty() || geom.back() != pnt)
+      geom.push_back(pnt);
+  }
+  TEST_EQUAL(geom, expectedRouteGeom, ());
+}
+
+void TestRestrictionPermutations(RestrictionVec restrictions, vector<m2::PointD> const & expectedRouteGeom,
+                                 RoadPoint const & start, RoadPoint const & finish,
+                                 RestrictionTest & restrictionTest)
+{
+  sort(restrictions.begin(), restrictions.end());
+  do
+  {
+    for (Restriction const & restriction : restrictions)
+    {
+      CHECK_EQUAL(restriction.m_featureIds.size(), 2, ());
+      RestrictionPoint restrictionPoint;
+      CHECK(restrictionTest. GetAdjacentFtPoint(restriction.m_featureIds[0], restriction.m_featureIds[1],
+                                                restrictionPoint), ());
+
+      switch (restriction.m_type)
+      {
+      case Restriction::Type::No:
+        restrictionTest.ApplyRestrictionNoRealFeatures(restrictionPoint);
+        continue;
+      case Restriction::Type::Only:
+        restrictionTest.ApplyRestrictionOnlyRealFeatures(restrictionPoint);
+        continue;
+      }
+    }
+    restrictionTest.SetStarter(start, finish);
+    TestRouteGeometry(*restrictionTest.m_starter, AStarAlgorithm<IndexGraphStarter>::Result::OK,
+                      expectedRouteGeom);
   }
+  while(next_permutation(restrictions.begin(), restrictions.end()));
```

Comment:
I guess it's important to have several tests which confirm that any permutation of restrictions let to generate route with the same geometry. But yes, there should be tests without permutations. 

Let's write the problematic permutation to log message.